### PR TITLE
Added pagination support for /user/teams github api with link header .

### PR DESCRIPTION
**Pull request by toshi-miura**
## Description

Added handling of last page pointed out below to [#102](https://github.com/pusher/oauth2_proxy/pull/102)
>Could we maybe use the Link values to make sure we only call up until the last page?

Test is done by rewriting the code with a smaller "per_page" value.

### CODE REVIEW

File: providers/github.go [hunk 1]
```diff
@@ -139,36 +140,95 @@ func (p *GitHubProvider) hasOrgAndTeam(accessToken string) (bool, error) {
...
+
+		if last == 0 {
+			// link header may not be obtained
+			// When paging is not required and all data can be retrieved with a single call
+
+			// Conditions for obtaining the link header.
+			// 1. When paging is required (Example: When the data size is 100 and the page size is 99 or less)
+			// 2. When it exceeds the paging frame (Example: When there is only 10 records but the second page is called with a page size of 100)
+
+			// link herder at not last page
```
Comment (JoelSpeed) [inline providers/github.go]: Typo I think
```suggestion
			// link header at not last page
```

File: providers/github.go [hunk 2]
```diff
@@ -139,36 +140,95 @@ func (p *GitHubProvider) hasOrgAndTeam(accessToken string) (bool, error) {
...
+			// <https://api.github.com/user/teams?page=1&per_page=100>; rel="prev", <https://api.github.com/user/teams?page=1&per_page=100>; rel="last", <https://api.github.com/user/teams?page=1&per_page=100>; rel="first"
+			// link herder at last page (doesn't exist last info)
```
Comment (JoelSpeed) [inline providers/github.go]: Typo again?
```suggestion
			// link header at last page (doesn't exist last info)
```

File: providers/github.go [hunk 3]
```diff
@@ -139,36 +140,95 @@ func (p *GitHubProvider) hasOrgAndTeam(accessToken string) (bool, error) {
...
+			// <https://api.github.com/user/teams?page=3&per_page=10>; rel="prev", <https://api.github.com/user/teams?page=1&per_page=10>; rel="first"
+
+			logger.Printf("1:endpoint :[%s]",endpoint.String())
```
Comment (JoelSpeed) [inline providers/github.go]: Not sure we need to log these? Could become very noisy in the logs as these would both be printed on every request

File: providers/github.go [hunk 4]
```diff
@@ -139,36 +140,95 @@ func (p *GitHubProvider) hasOrgAndTeam(accessToken string) (bool, error) {
...
+			logger.Printf("2:link header:[%s]",resp.Header.Get("Link"))		
+			link := resp.Header.Get("Link")
+			rep1 := regexp.MustCompile(`(?s).*\<https://api.github.com/user/teams\?page=(.)&per_page=[0-9]+\>; rel="last".*`)
+			i,converr:= strconv.Atoi(rep1.ReplaceAllString(link, "$1"))
+			if converr != nil{
+				logger.Printf("3:doesn't exist last page info. ")
+			}else{
+				last = i
+				logger.Printf("3:last page index  :[%d]", last)
```
Comment (JoelSpeed) [inline providers/github.go]: Potentially over-logging here

Comment (toshi-miura) [reply providers/github.go]: Thank you for the review. How about fixing it?

File: providers/github.go [hunk 5]
```diff
@@ -139,36 +140,90 @@ func (p *GitHubProvider) hasOrgAndTeam(accessToken string) (bool, error) {
...
+			// <https://api.github.com/user/teams?page=3&per_page=10>; rel="prev", <https://api.github.com/user/teams?page=1&per_page=10>; rel="first"
+
+
+			link := resp.Header.Get("Link")
+			rep1 := regexp.MustCompile(`(?s).*\<https://api.github.com/user/teams\?page=(.)&per_page=[0-9]+\>; rel="last".*`)
+			i,converr:= strconv.Atoi(rep1.ReplaceAllString(link, "$1"))
+			if converr != nil{
+			}else{
+				last = i
+			}
```
Comment (JoelSpeed) [inline providers/github.go]: I would suggest we  make this just
```
if converr == nil {
  last = i 
}
```
and have no else case, WDYT?

Comment (toshi-miura) [reply providers/github.go]: Thank you for the review.
Corrected as pointed out and added comments

**Comment by toshi-miura** [comment]
@JoelSpeed 

>Please rebase on master and add a note to the changelog for the entry and we can then get this merged!

My statement was rebase, except for changes to the original author.

I didn't edit CHANGELOG.md.
Since # 102 was taken over and the change difference of CHANGELOG.md was large

File: providers/github.go [hunk 6]
```diff
@@ -139,36 +140,89 @@ func (p *GitHubProvider) hasOrgAndTeam(accessToken string) (bool, error) {
...
+			// If the last page cannot be taken from the link in the http header, the last variable remains zero
+			if converr == nil {
+				last = i
+			}
+		}
+
+		resp.Body.Close()
+		if err != nil {
+			return false, err
+		}
```
Comment (syscll) [inline providers/github.go]: For readability, I would move this code directly after attempting to read `resp.Body` on L174

```
if err != nil {
        resp.Body.Close()
        return false, err
}
resp.Body.Close()
```

**Comment by JoelSpeed** [comment]
@toshi-miura I appreciate the diff was large, but you can add a new changelog comment now there's no merge to handle right? Since this is a functional change it needs a changelog entry before it can be merged

**Comment by toshi-miura** [comment]
@JoelSpeed 

> it needs a changelog entry 

Does it mean to add to CHANGELOG.md?
Is it like the following?

>Changes since v4.0.0
>#227 Add Keycloak provider (@Ofinka)
>#273 Support Go 1.13 (@dio)
>#274 XXXXXXXXXXXXX (@XXXX)

**Comment by JoelSpeed** [comment]
> Is it like the following?

Yep that's exactly what I'd like you to add, thank you :)

**Comment by toshi-miura** [comment]
Can't pull the latest master CHANGELOG.md because I am a child of # 102, is that okay?

**Comment by toshi-miura** [comment]
I would like to ask you what kind of cycle is it to make a release binary?
In my job, I'm wondering whether to use what I built or use this release build.

**Comment by toshi-miura** [comment]
I ’m doing a manual test.
In my environment, there are only about 50 github teams, so I have the following two types of tests

- Test with this source code (Paging logic doesn't work )
- It's bad, but test with editing source code ("per_page": {"5"}, (Paging logic  work ))

File: CHANGELOG.md [hunk 7]
```diff
@@ -5,6 +5,7 @@
 - [#227](https://github.com/pusher/oauth2_proxy/pull/227) Add Keycloak provider (@Ofinka)
 - [#273](https://github.com/pusher/oauth2_proxy/pull/273) Support Go 1.13 (@dio)
 - [#275](https://github.com/pusher/oauth2_proxy/pull/275) docker: build from debian buster (@syscll)
+- [#274](https://github.com/pusher/oauth2_proxy/pull/274)  Supports many github teams with api pagination support (@toshi-miura ,@apratina)
```
Comment (syscll) [inline CHANGELOG.md]:
```suggestion
- [#274](https://github.com/pusher/oauth2_proxy/pull/274)  Supports many github teams with api pagination support (@toshi-miura, @apratina)
```

Comment (toshi-miura) [reply CHANGELOG.md]: Thanks for the review, fixed

File: CHANGELOG.md [hunk 8]
```diff
@@ -5,6 +5,7 @@
 - [#227](https://github.com/pusher/oauth2_proxy/pull/227) Add Keycloak provider (@Ofinka)
 - [#273](https://github.com/pusher/oauth2_proxy/pull/273) Support Go 1.13 (@dio)
 - [#275](https://github.com/pusher/oauth2_proxy/pull/275) docker: build from debian buster (@syscll)
+- [#274](https://github.com/pusher/oauth2_proxy/pull/274) Supports many github teams with api pagination support (@toshi-miura,@apratina)
```
Comment (JoelSpeed) [inline CHANGELOG.md]: There should be a space after the comma ideally, sorry for being pedantic
```suggestion
- [#274](https://github.com/pusher/oauth2_proxy/pull/274) Supports many github teams with api pagination support (@toshi-miura, @apratina)
```

**Comment by toshi-miura** [comment]
I'm sorry to be late

**Comment by toshi-miura** [comment]
> Once the changelog conflict is resolved this should be good to merge.

I resolved the conflict.

**Comment by JoelSpeed** [comment]
I'm happy with this now, I've fixed up the changelog again, @syscll gonna override your requested changes as I can see this is now resolved

**Comment by toshi-miura** [comment]
@JoelSpeed 

Thank you for the merge.

If the release schedule is decided,
I would be grateful if you would tell the release schedule

**Comment by syscll** [comment]
We are planning to do a release in the next few weeks. We're just finalising a few other PRs.

**Comment by toshi-miura** [comment]
Thank you very mush!

**Pull request merged**